### PR TITLE
Add support for Razer BlackWidow V4 X

### LINKED ISSUE #2130: Support for Blackwidow V4 X
### Device Info

* Name: Razer Blackwidow V4 X
* Product Number: RZ03-04701900-R3U1
* Model Number: RZ03-0470

### Device Information
```
$ lsusb -d 1532: -v
Bus 001 Device 005: ID 1532:0293 Razer USA, Ltd Razer BlackWidow V4 X
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0293 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer BlackWidow V4 X
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x006d
    bNumInterfaces          4
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     157
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              2 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
```

```
$ ls -lah /dev/input/by-id/
total 0
drwxr-xr-x 2 root root 200 Sep 16 09:45 .
drwxr-xr-x 4 root root 600 Sep 16 09:45 ..
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Logitech_USB_Receiver-if01-event-mouse -> ../event2
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Logitech_USB_Receiver-if01-mouse -> ../mouse0
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Razer_Razer_BlackWidow_V4_X-event-if01 -> ../event5
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Razer_Razer_BlackWidow_V4_X-event-if03 -> ../event7
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Razer_Razer_BlackWidow_V4_X-event-kbd -> ../event3
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Razer_Razer_BlackWidow_V4_X-if01-event-kbd -> ../event4
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Razer_Razer_BlackWidow_V4_X-if02-event-mouse -> ../event6
lrwxrwxrwx 1 root root   9 Sep 16 09:45 usb-Razer_Razer_BlackWidow_V4_X-if02-mouse -> ../mouse1
```

<details>

```
(root)$ usbhid-dump -m 1532 -ed
001:005:003:DESCRIPTOR         1694903253.163770
 05 59 09 01 A1 01 85 01 09 02 A1 02 09 03 15 00
 27 FF FF 00 00 75 10 95 01 B1 03 09 04 09 05 09
 06 09 07 09 08 15 00 27 FF FF FF 7F 75 20 95 05
 B1 03 C0 85 02 09 20 A1 02 09 21 15 00 27 FF FF
 00 00 75 10 95 01 B1 02 C0 85 03 09 22 A1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 01 B1 02 09 23
 09 24 09 25 09 27 09 26 15 00 27 FF FF FF 7F 75
 20 95 05 B1 02 09 28 09 29 09 2A 09 2C 09 2D 15
 00 26 FF 00 75 08 95 05 B1 02 C0 85 04 09 50 A1
 02 09 03 09 55 15 00 25 08 75 08 95 02 B1 02 09
 21 15 00 27 FF FF 00 00 75 10 95 08 B1 02 09 51
 09 52 09 53 09 51 09 52 09 53 09 51 09 52 09 53
 09 51 09 52 09 53 09 51 09 52 09 53 09 51 09 52
 09 53 09 51 09 52 09 53 09 51 09 52 09 53 15 00
 26 FF 00 75 08 95 18 B1 02 C0 85 05 09 60 A1 02
 09 55 15 00 25 08 75 08 95 01 B1 02 09 61 09 62
 15 00 27 FF FF 00 00 75 10 95 02 B1 02 09 51 09
 52 09 53 15 00 26 FF 00 75 08 95 03 B1 02 C0 85
 06 09 70 A1 02 09 71 15 00 25 01 75 08 95 01 B1
 02 C0 85 07 06 00 FF 15 00 26 FF 00 09 02 75 08
 95 3F B1 02 C0 05 0C 09 01 A1 01 85 08 19 D0 29
 D7 15 00 25 01 75 01 95 08 81 02 C0

001:005:002:DESCRIPTOR         1694903253.167759
 05 01 09 02 A1 01 09 01 A1 00 05 09 19 01 29 05
 15 00 25 01 75 01 95 05 81 02 75 01 95 03 81 03
 06 00 FF 09 40 75 08 95 02 15 81 25 7F 81 02 05
 01 09 38 15 81 25 7F 75 08 95 01 81 06 09 30 09
 31 16 00 80 26 FF 7F 75 10 95 02 81 06 C0 06 00
 FF 09 02 15 00 25 01 75 08 95 5A B1 01 C0

001:005:001:DESCRIPTOR         1694903253.172763
 05 01 09 06 A1 01 85 01 05 07 19 E0 29 E7 15 00
 25 01 75 01 95 08 81 02 19 00 29 A0 15 00 25 01
 75 01 95 A0 81 02 C0 05 0C 09 01 A1 01 85 02 19
 00 2A 3C 02 15 00 26 3C 02 95 01 75 10 81 00 75
 08 95 13 81 01 C0 05 01 09 80 A1 01 85 03 19 81
 29 83 15 00 25 01 75 01 95 03 81 02 95 05 81 01
 75 08 95 14 81 01 C0 05 01 09 00 A1 01 85 04 09
 03 15 00 26 FF 00 35 00 46 FF 00 75 08 95 15 81
 00 C0 05 01 09 00 A1 01 85 05 09 03 15 00 26 FF
 00 35 00 46 FF 00 75 08 95 15 81 00 C0

001:005:000:DESCRIPTOR         1694903253.175763
 05 01 09 06 A1 01 05 07 19 E0 29 E7 15 00 25 01
 75 01 95 08 81 02 81 01 19 00 2A FF 00 15 00 26
 FF 00 75 08 95 06 81 00 05 08 19 01 29 03 15 00
 25 01 75 01 95 03 91 02 95 05 91 01 C0

```

</details>

Presumably it's relatively similar to Blackwidow V4 75% (already has an issue open), and Blackwidow V4 Pro (already has a PR open), but I thought it should be opened nonetheless.

### PR DESCRIPTION
**Pull request by tostivintc**
Hi,

I've successfully tested the RGB with polychromatic, but not yet the macro.

Fixes #2130

### CODE REVIEW

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -2001,6 +2013,7 @@ static ssize_t razer_attr_write_matrix_effect_static(struct device *dev, struct
     case USB_DEVICE_ID_RAZER_BLACKWIDOW_V3_MINI:
     case USB_DEVICE_ID_RAZER_BLADE_15_BASE_2022:
     case USB_DEVICE_ID_RAZER_TARTARUS_V2:
+//    case USB_DEVICE_ID_RAZER_BLACKWIDOW_V4_X:
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: If this is not needed, please remove the commented line.

**Comment by z3ntu** [comment]
> I've successfully tested the RGB with polychromatic, but not yet the macro.

If all the RGB works that's good!

Don't worry about the macro support, it's barely supported anyways and I'll probably remove the support completely at some point in favor of using e.g. https://github.com/sezanzeb/input-remapper

**Comment by squishypinkelephant** [comment]
`daemon/openrazer_daemon/hardware/keyboards.py` is missing `DEVICE_IMAGE` and will cause errors when using polychromatic. 
otherwise RGB works as expected.

**Comment by z3ntu** [comment]
@tostivintc Any update on this?

File: driver/razerkbd_driver.c (hunk 1 above)
Comment (tostivintc) [reply driver/razerkbd_driver.c]: It's now removed

**Comment by tostivintc** [comment]
I have added a device image link as requested and rebase on master head. But now I'm getting poll_rate error at polychromatic start. I need to dig to understand the issue.

**Comment by z3ntu** [comment]
What error do you get?

Btw the diff looks good now.

**Comment by tostivintc** [comment]
The error I got at polychromatic start:

```shell
Traceback (most recent call last):
  File ’/usr/bin/polychromatic-controller’, line 311, in _change_tab
    tab_objects[index].set_tab()
  File ’/usr/lib/python3/dist-packages/polychromatic/controller/devices.py’, line 90, in set_tab
    device_list = self.middleman.get_devices()
  File ’/usr/lib/python3/dist-packages/polychromatic/middleman.py’, line 151, in get_devices
    self._reload_device_cache_if_empty()
  File ’/usr/lib/python3/dist-packages/polychromatic/middleman.py’, line 122, in _reload_device_cache_if_empty
    device_list = backend.get_devices()
  File ’/usr/lib/python3/dist-packages/polychromatic/backends/openrazer.py’, line 153, in get_devices
    devices.append(self._get_device(rdevice))
  File ’/usr/lib/python3/dist-packages/polychromatic/backends/openrazer.py’, line 255, in _get_device
    main_zone.options.append(self._get_poll_rate_option(rdevice))
  File ’/usr/lib/python3/dist-packages/polychromatic/backends/openrazer.py’, line 1318, in _get_poll_rate_option
    current_rate = int(rdevice.poll_rate)
AttributeError: ’RazerKeyboard’ object has no attribute ’poll_rate’
``` 

And driver/dmesg error when a write is done on sysfs's poll_rate:
```shell
[175298.609439] razerkbd: Command not supported. status: 05 transaction_id.id: 1f remaining_packets: 00 protocol_type: 00 data_size: 02, command_class: 00, command_id.id: 40 Params: 00100000000000000000000000000000 .
```

**Comment by z3ntu** [comment]
Does Synapse on Windows allow you to the change the polling rate of the device? The firmware is saying the command is not supported so the device quite explicitly says this doesn't work.

**Comment by nfnvsc** [comment]
@z3ntu I just checked on Synapse (on Windows) and I cannot change the polling rate of the Blackwidow v4 X

**Comment by z3ntu** [comment]
Then remove polling rate support from this PR? Both driver and daemon.

**Comment by z3ntu** [comment]
Closing in favor of #2269, thanks for working on this!

**Pull request closed without merging**